man/man1/shmtcouplingmatrix.1
.\" Automatically generated by Pandoc 2.17.1.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "shmtcouplingmatrix" "1" "2021-02-15" "Fortran 95" "SHTOOLS 4.10"
.hy
.SH SHMTCouplingMatrix
.PP
This routine returns the multitaper coupling matrix for a given set of
power spectra of arbitrary localization windows.
This matrix relates the expectation of the localized multitaper spectrum
to the expectation of the power spectrum of the global function.
.SH Usage
.PP
call SHMTCouplingMatrix (\f[V]mmt\f[R],
\f[V]lmax\f[R],\f[V]tapers_power\f[R], \f[V]lwin\f[R], \f[V]k\f[R],
\f[V]taper_wt\f[R], \f[V]exitstatus\f[R])
.SH Parameters
.TP
\f[V]mmt\f[R] : output, real(dp), dimension (\f[V]lmax\f[R]+\f[V]lwin\f[R]+1, \f[V]lmax\f[R]+1)
The full multitaper coupling matrix that relates the expectation of the
localized multitaper spectrum to the global power spectrum of the
function.
.TP
\f[V]lmax\f[R] : input, integer(int32)
The spherical harmonic bandwidth of the global power spectrum.
.TP
\f[V]tapers_power\f[R] : input, real(dp), dimension (\f[V]lwin\f[R]+1, \f[V]k\f[R])
An array of power spectra of the k windowing functions, arranged in
columns.
.TP
\f[V]lwin\f[R] : input, integer(int32)
The spherical harmonic bandwidth of the windowing functions in the array
\f[V]tapers\f[R].
.TP
\f[V]k\f[R] : input, integer(int32)
The number of tapers utilized in the multitaper spectral analysis.
.TP
\f[V]taper_wt\f[R] : input, optional, real(dp), dimension (\f[V]k\f[R])
The weights used in calculating the multitaper spectral estimates.
Optimal values of the weights (for a known global power spectrum) can be
obtained from the routine \f[V]SHMTVarOpt\f[R].
.TP
\f[V]exitstatus\f[R] : output, optional, integer(int32)
If present, instead of executing a STOP when an error is encountered,
the variable exitstatus will be returned describing the error.
0 = No errors; 1 = Improper dimensions of input array; 2 = Improper
bounds for input variable; 3 = Error allocating memory; 4 = File IO
error.
.SH Description
.PP
\f[V]SHMTCouplingMatrix\f[R] returns the multitaper coupling matrix that
relates the expectation of the localized multitaper spectrum to the
expectation of the global power spectrum of the function (assumed to be
stationary).
This is given by eqs 4.5 and 4.6 in Wieczorek and Simons (2007):
.PP
\f[V]< S_{Phi Phi}\[ha](mt) > = M\[ha](mt) S_{ff}\f[R]
.PP
where \f[V]S_{Phi Phi}\f[R] is a vector containing the
\f[V]lmax+lwin+1\f[R] localized multitaper power spectral estiamtes,
\f[V]S_{ff}\f[R] is a vector of the global power spectrum up to degree
\f[V]lmax\f[R], and \f[V]< ... >\f[R] is the expectation operator.
The coupling matrix is given explicitly by
.PP
\f[V]M_{ij} = Sum_{l=0}\[ha]L Sum_{k=1}\[ha]K a_k S_{hh}\[ha]{k}(l) [ C_{l0j0}\[ha]{i0} ]\[ha]2\f[R]
.PP
where \f[V]a_k\f[R] are the taper weights, \f[V]S_{hh}\f[R] is the power
of the window, and \f[V]C\f[R] is a Clebsch-Gordon coefficient.
.PP
Note that this routine returns the \[lq]full\[rq] coupling matrix of
dimension (\f[V]lmax\f[R] + \f[V]lwin\f[R] + 1, \f[V]lmax\f[R] + 1).
When multiplied by a global input power spectrum with bandwidth
\f[V]lmax\f[R], it returns the output power spectrum with a bandwidth of
\f[V]lmax\f[R] + \f[V]lwin\f[R].
In doing so, it is implicitly assumed that input power spectrum is
exactly zero for all degrees greater than lmax.
If this is not the case, the ouput power spectrum should be considered
valid only for the degrees up to and including \f[V]lmax\f[R] -
\f[V]lwin\f[R].
.SH References
.PP
Wieczorek, M.
A.
and F.
J.
Simons, Minimum-variance multitaper spectral estimation on the sphere,
J.
Fourier Anal.
Appl., 13, 665-692, doi:10.1007/s00041-006-6904-1, 2007.
.SH See also
.PP
shmultitaperse, shmultitapercse, shreturntapers, shmtvaropt, shmtdebias


src/SHMTCouplingMatrix.f95
subroutine SHMTCouplingMatrix(Mmt, lmax, tapers_power, lwin, K, taper_wt, &
                              exitstatus)
!------------------------------------------------------------------------------
!
!   This routine returns the multitaper coupling matrix, which relates the
!   global input spectrum to the expectation of the localized multitaper
!   spectrum.
!
!   < S_{Phi Phi}^(mt) > = M^(mt) S_{ff}
!
!   This is given by eqs 4.5 and 4.6 of Wieczorek and Simons (2007).
!   The input tapers_power is a matrix containing the power spectra (in
!   columns) of all the localization windows. When using spherical cap
!   localization windows, this is simply
!
!   tapers_power[:,:] = tapers[:,:]**2
!
!   Note that this routine returns the "full" coupling matrix of dimension
!   (lmax+lwin+1, lmax+1). When multiplied by a global input power spectrum
!   with bandwidth lmax, it returns the output power spectrum with a bandwidth
!   of lmax+lwin. In doing so, it is implicitly assumed that input power
!   spectrum is exactly zero for all degrees greater than lmax. If this is not
!   the case, the ouput power spectrum should be considered valid only
!   for the degrees up to and including lmax-lwin.
!
!   Copyright (c) 2005-2019, SHTOOLS
!   All rights reserved.
!
!------------------------------------------------------------------------------
    use SHTOOLS, only: wigner3j
    use ftypes

    implicit none
    real(dp), intent(out) :: Mmt(:,:)
    real(dp), intent(in) :: tapers_power(:,:)
    real(dp), intent(in), optional :: taper_wt(:)
    integer(int32), intent(in) :: lmax, K, lwin
    integer(int32), intent(out), optional :: exitstatus
    real(dp) :: w3j(lwin+2*lmax+1), sum1
    integer(int32) :: i, j, l, wmin, wmax

    if (present(exitstatus)) exitstatus = 0

    if  (size(Mmt(:,1)) < lmax+lwin+1 .or. size(Mmt(1,:)) < lmax+1) then
        print*, "Error --- SHMTCouplingMatrix"
        print*, "MMT must be dimensioned as (LMAX+LWIN+1, LMAX+1) where "// &
                "LMAX and LWIN are ", lmax, lwin
        print*, "Input array is dimensioned as ", size(Mmt(:,1)), size(Mmt(1,:))
        if (present(exitstatus)) then
            exitstatus = 1
            return
        else
            stop
        end if

    end if

    if (size(tapers_power(:,1)) < lwin+1 .or. size(tapers_power(1,:)) < K) then
        print*, "Error --- SHMTCouplingMatrix"
        print*, "TAPERS_POWER must be dimensioned as (LWIN+1, K) where LWIN "// &
                "and K are ", lwin, k
        print*, "Input array is dimensioned as ", size(tapers_power(:,1)), &
            size(tapers_power(1,:))
        if (present(exitstatus)) then
            exitstatus = 1
            return
        else
            stop
        end if

    end if

    if (present(taper_wt)) then
        if (size(taper_wt) < K) then
            print*, "Error --- SHMTCouplingMatrix"
            print*, "TAPER_WT must be dimensioned as (K) where K is ", k
            print*, "Input array is dimensioned as ", size(taper_wt)
            if (present(exitstatus)) then
                exitstatus = 1
                return
            else
                stop
            end if
        end if

    end if

    !--------------------------------------------------------------------------
    !
    !   Compute coupling matrix, M^(mt)
    !
    !--------------------------------------------------------------------------

    if (present(taper_wt)) then

        do i=0, lmax+lwin
            do j=0, lmax
                if (present(exitstatus)) then
                    call Wigner3j(w3j, wmin, wmax, i, j, 0, 0, 0, &
                                  exitstatus = exitstatus)
                    if (exitstatus /= 0) return
                else
                    call Wigner3j(w3j, wmin, wmax, i, j, 0, 0, 0)
                end if
                sum1 = 0.0_dp

                do l = wmin, min(wmax,lwin), 2
                    sum1 = sum1 + dot_product(taper_wt(1:K), &
                           tapers_power(l+1,1:K)) * w3j(l-wmin+1)**2
                end do

                Mmt(i+1,j+1) = sum1 * dble(2*i+1)

            end do
        end do

    else

        do i = 0, lmax+lwin
            do j = 0, lmax
                if (present(exitstatus)) then
                    call Wigner3j(w3j, wmin, wmax, i, j, 0, 0, 0, &
                                  exitstatus = exitstatus)
                    if (exitstatus /= 0) return
                else
                    call Wigner3j(w3j, wmin, wmax, i, j, 0, 0, 0)
                end if
                sum1 = 0.0_dp

                do l = wmin, min(wmax,lwin), 2
                    sum1 = sum1 + sum(tapers_power(l+1,1:K)) * w3j(l-wmin+1)**2
                end do

                Mmt(i+1,j+1) = sum1 * dble(2*i+1) / dble(K)

            end do
        end do

    end if

end subroutine SHMTCouplingMatrix
